src/librustc_resolve/late.rs

```diff
@@ -544,6 +544,60 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
             self.visit_where_predicate(p);
         }
     }
+
+    fn visit_generic_arg(&mut self, arg: &'tcx GenericArg) {
+        debug!("visit_generic_arg({:?})", arg);
+        match arg {
+            GenericArg::Type(ref ty) => {
+                // We parse const arguments as path types as we cannot distiguish them durring
+                // parsing. We try to resolve that ambiguity by attempting resolution the type
+                // namespace first, and if that fails we try again in the value namespace. If
+                // resolution in the value namespace succeeds, we have an generic const argument on
+                // our hands.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    // Const arguments cannot have a multi-segment path right now, so we check that
+                    // first.
+                    if path.segments.len() == 1 {
+                        if let None = self.resolve_qpath(
+                            ty.id,
+                            qself.as_ref(),
+                            &Segment::from_path(path),
+                            Namespace::TypeNS,
+                            ty.span,
+                            CrateLint::SimplePath(ty.id),
+                        ) {
+                            if let Some(binding) = self.resolve_ident_in_lexical_scope(
+                                path.segments[0].ident,
+                                Namespace::ValueNS,
+                                Some(ty.id),
+                                ty.span,
+                            ) {
+                                debug!(
+                                    "visit_generic_arg: generic type argument resolved as a const"
+                                );
+                                self.with_constant_rib(|this| {
```

Comment:
`with_constant_rib` is supposed to be opened before we resolve the path in value namespace, not after.
This can be tested by attempting to use something non-const, like a local variable, in a generic argument.

Reply:
In general, this resolution in value namespace should reuse code from `visit_anon_const` as much as possible instead of adding its own subtly different code path.
This means `with_constant_rib` on the outside, then copypaste of the `ExprKind::Path` case from `resolve_expr`.
(Unfortunately, we cannot use `visit_anon_const` directly here due to visitor lifetimes.)

Reply:
I have attempted to wrap the `with_constant_rib` around the resolution in the value namespace, but have found it not to work quite right.

If I use `smart_resolve_path` as done in the `ExprKind::Path` branch of `resolve_expr` for the value namespace resolution, core doesn't even compile. This is due to macros breaking as all generic args in macros start to be seen as values, even where they should be types. I had avoided `smart_resolve_path` to begin with because that smartness includes throwing a bunch of errors that would be erroneous in this case and it doesn't provide any feedback about success to the caller.

If I use `resolve_ident_in_lexical_scope` for the value namespace resolution everything compiles fine, except the macros in the test "ui/macros/type-macros-simple.rs". In this case the constant rib starts complaining that the primitive types are not constant values.

Now, this is all with using `resolve_ident_in_lexical_scope` for type value resolution. If `resolve_qpath` is used instead, as it is done currently in the patches, everything is fine when using `smart_resolve_path` for value namespace resolution, though there are quite a few error message changes seen in tests. I have pushed these changes up for everyone to review.

Reply:
Ok, I'll look what happens when `resolve_ident_in_lexical_scope` is used for the speculative check in type namespace, because what you describe sounds strange.

>when using smart_resolve_path for value namespace resolution, though there are quite a few error message changes seen in tests.

Yeah, we'll probably have to use two speculative checks to improve the diagnostics for the case when the name cannot be found in both namespaces.
Basically, side-effectful `smart_resolve_path` for values should not be called if a "pure" speculative check fails in value namespace. Then we'll still get "unresolved type" errors and not "unresolved value".

Reply:
Here's the fix - https://github.com/petrochenkov/rust/commit/7c3a4027e9f527151297e6c70b54e67885ebce47.
`resolve_ident_in_lexical_scope` had an issue with built-in types that was generally masked in `resolve_qpath` a couple of levels above, but still prevented speculative use of `resolve_ident_in_lexical_scope` in type namespace.